File: javascript/binary-search.js
function binarySearch(arr, target) {
    let left = 0;
    let right = arr.length-1;

    // the main idea of binary search is to cut search area by adjust left or right position 
    // and check at middle position of the area. array must be sorted!
    while(left <= right) {
        let middle = Math.floor((left+right)/2);
        if (arr[middle] == target) {
            return middle;
        }
        // adjust area to left side by change right point.
        else if (arr[middle] > target) {
            right = middle-1;
        }
        // adjust area to right side by change left point. 
        else {
            left = middle+1;
        }
    }
    return null;
}

console.log("index", binarySearch([1,2,3,4,5,6,7,8,9], 8));

File: golang/binary-search/binary-search.go
package main

import (
	"fmt"
	"math"
)

func binarySearch(arr []int, target int) int {
	left := 0
	right := len(arr) - 1

	// the main idea of binary search is to cut search area by adjust left or right position
	// and check at middle position of the area. array must be sorted!
	for left <= right {
		middle := int(math.Ceil(float64(left+right) / float64(2)))
		// if match then return the array index
		if arr[middle] == target {
			return middle
		} else if arr[middle] > target { // adjust area to left side by change right point.
			right = middle - 1
		} else { // adjust area to right side by change left point.
			left = middle + 1
		}
	}
	return -1
}

func main() {
	result := binarySearch([]int{1, 2, 3, 4, 6, 7, 8, 9, 10}, 8)
	fmt.Println("result", result)
}
